homeassistant/components/hyperion/__init__.py

```diff
@@ -151,23 +197,86 @@ async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> b
+            return
+        await async_instances_to_clients_raw(response[hyperion_const.KEY_DATA])
+
+    async def async_instances_to_clients_raw(instances: List[Dict[str, Any]]) -> None:
+        """Convert instances to Hyperion clients."""
+        registry = await async_get_registry(hass)
+        running_instances: Set[int] = set()
+        stopped_instances: Set[int] = set()
+        existing_instances = hass.data[DOMAIN][config_entry.entry_id][
+            CONF_INSTANCE_CLIENTS
+        ]
+        server_id = cast(str, config_entry.unique_id)
+
+        # In practice, an instance can be in 3 states as seen by this function:
+        #
+        #    * Exists, and is running: Should be present in HASS/registry.
+        #    * Exists, but is not running: Cannot add it yet, but entity may have be
+        #      registered from a previous time it was running.
+        #    * No longer exists at all: Should not be present in HASS/registry.
+
+        # Add instances that are missing.
+        for instance in instances:
+            instance_num = instance.get(hyperion_const.KEY_INSTANCE)
+            if instance_num is None:
+                continue
+            if not instance.get(hyperion_const.KEY_RUNNING, False):
+                stopped_instances.add(instance_num)
+                continue
+            running_instances.add(instance_num)
+            if instance_num in existing_instances:
+                continue
+            hyperion_client = await async_create_connect_hyperion_client(
+                host, port, instance=instance_num, token=token
+            )
+            if not hyperion_client:
+                continue
+            existing_instances[instance_num] = hyperion_client
+            instance_name = instance.get(hyperion_const.KEY_FRIENDLY_NAME, DEFAULT_NAME)
+            async_dispatcher_send(
+                hass,
+                SIGNAL_INSTANCE_ADD.format(config_entry.entry_id),
+                instance_num,
+                instance_name,
+            )
+
+        # Remove entities that are are not running instances on Hyperion.
+        for instance_num in set(existing_instances) - running_instances:
+            del existing_instances[instance_num]
+            async_dispatcher_send(
+                hass, SIGNAL_INSTANCE_REMOVE.format(config_entry.entry_id), instance_num
+            )
+
+        # Deregister entities that belong to removed instances.
+        for entry in async_entries_for_config_entry(registry, config_entry.entry_id):
+            data = split_hyperion_unique_id(entry.unique_id)
+            if not data:
+                continue
+            if data[0] == server_id and data[1] not in running_instances.union(
+                stopped_instances
+            ):
+                registry.async_remove(entry.entity_id)
+
+    hyperion_client.set_callbacks(
+        {
+            f"{hyperion_const.KEY_INSTANCE}-{hyperion_const.KEY_UPDATE}": async_instances_to_clients,
```

Comment:
fyi you can also use tuples as keys in dictionaries:
```suggestion
            (hyperion_const.KEY_INSTANCE, hyperion_const.KEY_UPDATE): async_instances_to_clients,
```

Reply:
Thanks! The "-update" syntax is actually direct from Hyperion-NG itself and just piped through the Python library [ref](https://docs.hyperion-project.org/en/json/Subscribe.html#base-response-layout).